### DISCUSSION TITLE
Create a packfile from objects (without a repository)

### DISCUSSION THREAD

**Discussion by swedgwood** [GitoxideLabs/gitoxide#831] [Q&A] [answered]
Hi!

I'm writing a toy HTTP-based Git remote, and using gitoxide here and there for some plumbing

I was just wondering if it is possible to create a packfile (suitable for using in a response to a fetch command when cloning) from a list of objects, each just having its object ID and the raw object data (i.e. something like `Vec<(String, Vec<u8>)>`).

I've read around in `gix-pack`, but I can't figure out how I might do this.

I've also looked at https://github.com/Byron/gitoxide/discussions/692 and read through [this create function](https://github.com/Byron/gitoxide/blob/3c4e0b4997e86aa85d303da188f5eb55905bc1f2/gitoxide-core/src/pack/create.rs#L99), but so much of the code seems to rely on a handle to an actual repository. Is there an easy way to do this when I simply have the raw object data and the IDs?

Thanks in advance!

**Comment by Byron** [comment, marked as answer]
Pack creation is indeed possible and abstracted enough to be able to create packs entirely from memory.

To get started, I recommend playing with `gix no-repo pack`, and the implementation [can be found here](https://github.com/Byron/gitoxide/blob/14e64e74649cfb1f2f99da87015939af98fae5c8/gitoxide-core/src/pack/create.rs#L93) for prototyping.

But you already know that.

> I've also looked at https://github.com/Byron/gitoxide/discussions/692 and read through [this create function](https://github.com/Byron/gitoxide/blob/3c4e0b4997e86aa85d303da188f5eb55905bc1f2/gitoxide-core/src/pack/create.rs#L99), but so much of the code seems to rely on a handle to an actual repository. Is there an easy way to do this when I simply have the raw object data and the IDs?

The `handle` in this portion of the codebase refers to something that implements the `gix_pack::Find` trait. You should be able to implement portions of it to provide data to the machinery. Note that it's tuned to assume packs, but it can work without that as not every object is packed.

I hope that helps.
